src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -123,94 +150,119 @@ impl DatasetRepository for DatasetRepositoryS3 {
-                        .await
-                        .map_err(|e| {
-                            if let GetSummaryError::EmptyDataset = e {
-                                GetDatasetError::NotFound(DatasetNotFoundError {
-                                    dataset_ref: dataset_ref.clone(),
-                                })
-                            } else {
-                                GetDatasetError::Internal(e.int_err())
-                            }
-                        })?;
-
-                    Ok(DatasetHandle::new(summary.id, alias.clone()))
-                } else {
-                    Err(GetDatasetError::NotFound(DatasetNotFoundError {
-                        dataset_ref: dataset_ref.clone(),
-                    }))
-                }
-            }
-            DatasetRef::ID(id) => {
                 // TODO: this is really really slow and expensive!
+                let normalized_alias = self.normalize_alias(alias);
                 use futures::StreamExt;
                 let mut datasets = self.get_all_datasets();
                 while let Some(hdl) = datasets.next().await {
                     let hdl = hdl?;
-                    if hdl.id == *id {
+                    if hdl.alias == normalized_alias {
                         return Ok(hdl);
                     }
                 }
                 Err(GetDatasetError::NotFound(DatasetNotFoundError {
                     dataset_ref: dataset_ref.clone(),
                 }))
             }
+            DatasetRef::ID(id) => {
+                if self
+                    .s3_context
+                    .bucket_path_exists(&id.cid.to_string())
+                    .await?
+                {
+                    let dataset = self.get_dataset_impl(id).await?;
+                    let dataset_alias = self.resolve_dataset_alias(&dataset).await?;
+                    Ok(DatasetHandle::new(id.clone(), dataset_alias))
+                } else {
+                    Err(GetDatasetError::NotFound(DatasetNotFoundError {
+                        dataset_ref: dataset_ref.clone(),
+                    }))
+                }
+            }
         }
     }
 
     fn get_all_datasets<'s>(&'s self) -> DatasetHandleStream<'s> {
-        Box::pin(async_stream::try_stream! {
-            let folders_common_prefixes = self.s3_context.bucket_list_folders().await?;
-            for prefix in folders_common_prefixes {
-                let mut prefix = prefix.prefix.unwrap();
-                while prefix.ends_with('/') {
-                    prefix.pop();
-                }
-
-                let name = DatasetAlias::new(None, DatasetName::try_from(prefix).int_err()?);
-                let hdl = self.resolve_dataset_ref(&name.into()).await.int_err()?;
-                yield hdl;
-            }
-        })
+        self.stream_datasets_if(|_| true)
     }
 
     fn get_account_datasets<'s>(&'s self, account_name: AccountName) -> DatasetHandleStream<'s> {
```

Comment:
Should we rename this method to `get_datasets_by_owner`? It's a bit ambiguous about the relationship it's checking:
- is it all datasets an account has access to
- all datasets it owns
- or something else.

Reply:
Right, clarification makes sense